### PR TITLE
[pt2] add meta function for `solve_triangular`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #100833
* __->__ #100829

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -315,6 +320,51 @@ def squareCheckInputs(self: Tensor, f_name: str):
...
+        A.size(-1) == self.size(-2),
+        lambda: (
+            f"Incompatible matrix sizes for {name}: each A "
+            f"matrix is {A.size(-1)} by {A.size(-1)}"
+            f" but each b matrix is ",
+            self.size(-2),
+            " by ",
+            self.size(-1),
+        ),
+    )
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:365]: Need to fix this error msg.

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -483,6 +551,62 @@ def _linalg_svd_meta(
     return U, S, V
 
 
+def _linalg_broadcast_batch_dims(arg1, arg2):
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:554]: Add types here.

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -483,6 +551,62 @@ def _linalg_svd_meta(
...
+
+    arg1_expand_size = list(expand_batch_portion)
+    arg1_expand_size += [arg1.size(-2), arg1.size(-1)]
+
+    arg2_expand_size = list(expand_batch_portion)
+    arg2_expand_size += [arg2.size(-2), arg2.size(-1)]
+    return arg1_expand_size, arg2_expand_size
+
+
+def _linalg_broadcast_batch_dims_name(arg1, arg2, name):
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:568]: Add types here.

File: torch/_prims_common/wrappers.py [hunk 4]
```diff
@@ -139,22 +139,29 @@ def _fn(*args, **kwargs):
         return _fn
 
 
-# TODO: handle tuples of tensors
-def _maybe_resize_out(out: TensorLikeType, shape: ShapeType):
+# Returns true if resize is necessary
```
Comment (nkaretnikov) [inline torch/_prims_common/wrappers.py:142]: Capitalize true.

File: torch/_meta_registrations.py [hunk 5]
```diff
@@ -315,6 +320,48 @@ def squareCheckInputs(self: Tensor, f_name: str):
...
+    )
+
+    check(
+        A.size(-1) == self.size(-2),
+        lambda: (
+            f"Incompatible matrix sizes for {name}: each A "
+            f"matrix is {A.size(-1)} by {A.size(-1)}"
+            f" but each b matrix is {self.size(-2)} by {self.size(-1)}"
+        ),
+    )
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:362]: See `linearSolveCheckInputs` in core.

File: torch/_meta_registrations.py [hunk 6]
```diff
@@ -339,6 +386,24 @@ def checkIsMatrix(A: Tensor, f_name: str, arg_name: str = "A"):
...
+    squareCheckInputs(A, f_name)
+    checkIsMatrix(B, f_name)
+    check(
+        A.size(-2) == B.size(-2) if left else A.size(-1) == B.size(-1),
+        lambda: (
+            f"{f_name}: Incompatible shapes of A and B for the equation "
+            f"{'AX = B' if left else 'XA = B'}"
+            f" ({A.size(-2)}x{A.size(-1)} and {B.size(-2)}x{B.size(-1)})"
+        ),
+    )
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:404]: See `checkInputsSolver` in core.

File: torch/_meta_registrations.py [hunk 7]
```diff
@@ -483,6 +548,62 @@ def _linalg_svd_meta(
...
+    arg1_batch_sizes = arg1.shape[:-2]
+    arg2_batch_sizes = arg2.shape[:-2]
+    expand_batch_portion = _broadcast_shapes(arg1_batch_sizes, arg2_batch_sizes)
+
+    arg1_expand_size = list(expand_batch_portion)
+    arg1_expand_size += [arg1.size(-2), arg1.size(-1)]
+
+    arg2_expand_size = list(expand_batch_portion)
+    arg2_expand_size += [arg2.size(-2), arg2.size(-1)]
+    return arg1_expand_size, arg2_expand_size
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:562]: See `_linalg_broadcast_batch_dims` in core (the overload with two params).

File: torch/_meta_registrations.py [hunk 8]
```diff
@@ -483,6 +548,62 @@ def _linalg_svd_meta(
...
+
+    arg1_expand_size, arg2_expand_size = _linalg_broadcast_batch_dims(arg1, arg2)
+
+    arg1_broadcasted = (
+        arg1 if arg1_expand_size == arg1.shape else arg1.expand(arg1_expand_size)
+    )
+    arg2_broadcasted = (
+        arg2 if arg2_expand_size == arg2.shape else arg2.expand(arg2_expand_size)
+    )
+    return arg1_broadcasted, arg2_broadcasted
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:578]: See `_linalg_broadcast_batch_dims` in core (the overload with three params, one string param).

File: torch/_meta_registrations.py [hunk 9]
```diff
@@ -483,6 +548,62 @@ def _linalg_svd_meta(
...
+    B_, A_ = _linalg_broadcast_batch_dims_name(B, A, None)
+    avoid_copy_A = A_.transpose(-2, -1).is_contiguous() and A_.is_conj()
+    if avoid_copy_A:
+        out = _maybe_resize_out(out, B_.shape)
+    else:
+        # reimplementation of resize_output with result F-contig
+        if _resize_output_check(out, B_.shape):
+            out.resize_(B_.transpose(-2, -1).shape)
+            out.transpose_(-2, -1)
+    return out  # type: ignore[return-value]
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:604]: See `linalg_solve_triangular` and `linalg_solve_triangular_out` in core.

File: torch/_prims_common/wrappers.py [hunk 10]
```diff
@@ -139,22 +139,29 @@ def _fn(*args, **kwargs):
...
+    if out.numel() != 0:
+        msg = (
+            f"An output with one or more elements was resized since it had shape {str(out.shape)} "
+            "which does not match the required output shape {str(shape)}. "
+            "This behavior is deprecated, and in a future PyTorch release outputs will not "
+            "be resized unless they have zero elements. "
+            "You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0)."
+        )
+        warnings.warn(msg)
+    return True
```
Comment (nkaretnikov) [inline torch/_prims_common/wrappers.py:156]: Meta func calls `_resize_output_check`, so I've split it out.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**